pkg/ebpf/events_pipeline.go

```diff
@@ -281,75 +286,100 @@ func (t *Tracee) decodeEvents(outerCtx context.Context, sourceChan chan []byte)
 	return out, errc
 }
 
-// computePolicies iterates through the policies that do the filtering in user space, checking whether an event should be considered.
-// If it should not, it sets the respective offset to 0.
-// Finally it returns computed policies and the names of the policies that matched.
-func (t *Tracee) computePolicies(event *trace.Event) (uint64, []string) {
+// matchPolicies does the userland filtering (policy matching) for events. It iterates through all
+// existing policies, that were set by the kernel in the event bitmask. Some of those policies might
+// not match the event after userland filters are applied. In those cases, the policy bit is cleared
+// (so the event is "filtered" for that policy).
+func (t *Tracee) matchPolicies(event *trace.Event) uint64 {
 	eventID := events.ID(event.EventID)
-	matchedPolicies := event.MatchedPolicies
-	matchedPoliciesNames := []string{}
+	bitmap := event.MatchedPoliciesKernel
 
-	for p := range t.config.Policies.Map() {
+	for p := range t.config.Policies.Map() { // range through each existing policy
+		// Policy ID is the bit offset in the bitmap.
 		bitOffset := uint(p.ID)
 
-		// Events submitted with matching policies.
-		// The policy must have its bit cleared when it does not match.
-		if !utils.HasBit(matchedPolicies, bitOffset) {
+		if !utils.HasBit(bitmap, bitOffset) { // event does not match this policy
 			continue
 		}
 
+		// The event might have this policy bit set, but the policy might not have this event ID.
+		// This happens whenever the event submitted by the kernel is going to derive an event that
+		// this policy is interested in. In this case, don't do anything and let the derivation
+		// stage handle this event.
+		//
+		// TODO: should the base event set policy bit for policies that depend on their derivative
+		//       events? (check decodeEvents)
```

Comment:
Oh, I think I got your enquires. But which base events depends on their derivatives? Wouldn't it be the other way around?

Reply:
We would have to know before hand what events that base event would be capable of deriving. Then we could mark the matched policies with all of them. But the thing is, should we do this ?

Because if we agree we should (For all base events) then only ONE of the //TODO marks would continue to exist.